### PR TITLE
N-step GRU and Bi-GRU of Cudnn [WIP]

### PR DESCRIPTION
**Pull request by aonotas**
Work in progress to support other types of RNNs using cuDNN 5. See Issue #1843.

Thanks @honnibal and @ywatanabex !

TODO:
- [x] N-step GRU
- [x] N-step BiGRU
- [x] N-step LSTM
- [ ] N-step BiLSTM
- [ ] N-step RNN-Relu
- [ ] N-step BiRNN-Relu
- [ ] N-step RNN-Tanh
- [ ] N-step BiRNN-Tanh
- [ ] Write documentation
- [ ]  Support float16, float64

### CODE REVIEW

**Comment by delta2323** [comment]
Thank you for the PR! As the PR has been large and will be larger, is it possible to split the PR into smaller ones (e.g. create a PR that implements NStep LSTM, first). Thanks in advance.

**Comment by delta2323** [comment]
Do you have an idea how NStep (Bi)GRU and (Bi)LSTM behave when cuDNN is not available?

File: chainer/functions/connection/n_step_gru.py [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+
+from chainer import cuda
+from chainer.functions.activation import sigmoid
+from chainer.functions.activation import tanh
+from chainer.functions.array import concat
+from chainer.functions.array import reshape
+from chainer.functions.array import split_axis
+from chainer.functions.array import stack
+from chainer.functions.connection import linear
+from chainer.functions.connection.n_step_lstm import NStepRNN
```
Comment (delta2323) [inline chainer/functions/connection/n_step_gru.py]: direct import of objects are prohibited in pep8

File: chainer/functions/connection/n_step_lstm.py [hunk 2]
```diff
@@ -117,33 +117,92 @@ def _split(inputs, pos):
...
+        libcudnn.CUDNN_RNN_RELU: {'n_W': 2, 'n_Wb': 2 * 2, 'n_cell': 1},
+        libcudnn.CUDNN_RNN_TANH: {'n_W': 2, 'n_Wb': 2 * 2, 'n_cell': 1},
+        libcudnn.CUDNN_GRU: {'n_W': 6, 'n_Wb': 2 * 6, 'n_cell': 1},
+        libcudnn.CUDNN_LSTM: {'n_W': 8, 'n_Wb': 2 * 8, 'n_cell': 2},
+        libcudnn.CUDNN_UNIDIRECTIONAL: {'n': 1},
+        libcudnn.CUDNN_BIDIRECTIONAL: {'n': 2}
+    }
+
+
+class NStepRNN(function.Function):
```
Comment (delta2323) [inline chainer/functions/connection/n_step_lstm.py]: Is it possible to remove `NStepRNN` from `n_step_lstm.py` and create new file for it, as `NStepGRU` and `NStepBiGRU` also use it?

Comment (aonotas) [inline chainer/functions/connection/n_step_lstm.py]: Thank you for your reply!
I will create other small PR.

Comment (aonotas) [inline chainer/functions/connection/n_step_lstm.py]: I create smaller pull-request.
https://github.com/pfnet/chainer/pull/1943

File: chainer/links/connection/n_step_gru.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    """Uni-directional GRU links
+
+    MEMO (watanabe):
+        Bidirectional GRU consits of several GRUs (w.r.t layers, directions)
+        each GRU can be accessed by `bigru[lid]`
+        where lid = (2 * layer + di)
+    """
+    def __init__(self, n_layers, in_size, out_size, dropout, use_cudnn=True):
+        NStepGRUBase.__init__(self, n_layers, in_size, out_size, dropout,
+                              direction='uni', use_cudnn=True)
```
Comment (ShigekiKarita) [inline chainer/links/connection/n_step_gru.py]: arg of `use_cudnn` should be used?: `use_cudnn=use_cudnn`

File: chainer/links/connection/n_step_gru.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+    MEMO (watanabe):
+        Bidirectional GRU consits of several GRUs (w.r.t layers, directions)
+        each GRU can be accessed by `bigru[lid]`
+        where lid = (2 * layer + di)
+    """
+    def __init__(
+            self, n_layers, in_size, out_size, dropout, use_cudnn=True):
+        NStepGRUBase.__init__(self, n_layers, in_size, out_size, dropout,
+                              direction='bi', use_cudnn=True)
```
Comment (ShigekiKarita) [inline chainer/links/connection/n_step_gru.py]: arg of `use_cudnn` should be used?: `use_cudnn=use_cudnn`

**Comment by aonotas** [comment]
Thank you for your comment.
I will fix it!

**Pull request closed without merging**